### DISCUSSION TITLE
Integration with neotest

### DISCUSSION THREAD

**Discussion by nikensss** [mrcjkb/rustaceanvim#382] [Q&A] [answered]
Hi! I'm trying to run tests in my project using neotest (which I managed to get to to work for TS projects), but I keep getting an error that it's not really clear to me were should I start looking. Sorry if this is not a problem with rustaceanvim.

The error that I get is really short and without much information: `error: failed to parse test binary arguments `--show-output`: arguments are unsupported`.

When I do `RustLsp debuggables` I can see some element of the list showing the `--show-output` flag. And everything works fine.

Any help appreciated, and thanks for the plugin, it's extremely useful.

**Comment by mrcjkb** [comment, marked as answer]
**Update: This has been fixed: https://github.com/mrcjkb/rustaceanvim/pull/384**

Hey 👋 

It looks like the executable you are using to run tests doesn't support the `--show-output` flag (which is provided by rust-analyzer).
rustaceanvim will use cargo-nextest if you have it installed.
You can try disabling that by setting `vim.g.rustaceanvim.tools.enable_nextest = false`.

If that fixes it, then it's probably a bug in rustaceanvim, and the command should probably be sanitised for cargo-nextest.

**Reply by nikensss** [reply]
Yup, that fixes the problem. I also don't really know why I have `nextest`, to be honest, I'm gonna do some research there. 

Thank you very much for the quick reply also!

**Reply by mrcjkb** [reply]
Thanks for the info 🙏

I've created an issue for it and will fix it when I'm near my laptop https://github.com/mrcjkb/rustaceanvim/issues/383

**Comment by PMassicotte** [comment]
Hi.

I am trying to use this feature with `neotest`. I am currently learning rust and doing some exercises. In these, there is a test file. For example:

![image](https://github.com/mrcjkb/rustaceanvim/assets/4519221/32fc5d11-d6cc-45c8-ae2f-39a20b5e433b)

As you can see, some tests are failing but reported as success when I run ` :lua require("neotest").run.run(vim.fn.expand("%"))` (see `perfect_lower_case`).

So it looks like that only the first test is run (`empty_sentence`), and all others marked `#[ignore]` are not run but marked success.

Another question, is it possible to run all test a la `cargo test -- --include-ignored`?

**Reply by mrcjkb** [reply]
Passing test args to the neotest adapter isn't supported yet: https://github.com/mrcjkb/rustaceanvim/issues/445

**Reply by PMassicotte** [reply]
Thank you for your answer. So at the moment I removed all ignore tags. One more question if you do not mind. I am now trying to run all tests, but I am getting a symbol that looks like that all but 1 test were not executed despite using `:lua require("neotest").run.run(vim.fn.expand("%"))` 

![Peek 2024-07-09 11-33](https://github.com/mrcjkb/rustaceanvim/assets/4519221/6757880a-faff-41e4-98b0-ebf37cdbd12a)

**Reply by mrcjkb** [reply]
Cargo doesn't have a concept of a "test file", so the neotest adapter will look for a top-level test module and run the first one it finds (or all of them - I don't remember).

In Rust, the convention is to put your tests in a separate test module: 

```rust
#[cfg(test)]
mod tests {

    use super::*;
   
    #[test]
    fn fooTest() {
    }
}
```

This has the additional benefit that it won't be included in your production binary.
